gcp_variant_transforms/libs/bq_to_vcf_temp_files_composer.py

```diff
@@ -0,0 +1,144 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Composes multiple files in GCS to one VCF file."""
+
+import multiprocessing
+from typing import Iterable, List  # pylint: disable=unused-import
+
+from apache_beam.io import filesystems
+from apache_beam.io.gcp import gcsio
+
+from google.cloud import storage
+
+_MAX_NUM_OF_BLOBS_PER_COMPOSE = 32
+
+
+def compose_temp_files(project, output_file, bq_to_vcf_temp_folder):
+  # type: (str, str, str) -> None
+  """Composes intermediate files to one VCF file.
+
+  It composes VCF data shards to one VCF data file and deletes the intermediate
+  VCF shards.
+  TODO(allieychen): Eventually, it further consolidates the meta information,
+  data header line, and the composed VCF data file into the `output_file`.
+
+  Args:
+    project: The project name.
+    output_file: The final VCF file path.
+    bq_to_vcf_temp_folder: The folder that contains all BigQuery to VCF pipeline
+     temp files.
+  """
+  bucket_name, blob_prefix = gcsio.parse_gcs_path(bq_to_vcf_temp_folder)
+  client = storage.Client(project)
+  bucket = client.get_bucket(bucket_name)
+  multi_process_composer = MultiProcessComposer(project, bucket_name,
+                                                blob_prefix)
+  composed_vcf_data_file = multi_process_composer.run()
+  output_file_blob = _create_blob(client, output_file)
+  output_file_blob.rewrite(composed_vcf_data_file)
+  bucket.delete_blobs(bucket.list_blobs(prefix=blob_prefix))
+
+
+def _compose_files(project, bucket_name, blob_names, composite_name):
+  # type: (str, str, List[str], str) -> None
+  """Composes multiple files (up to 32 objects) in GCS to one.
+
+  Args:
+    project: The project name.
+    bucket_name: The name of the bucket where the `components` and the new
+      composite are saved.
+    blob_names: A list of blob object names.
+    composite_name: Name of the new composite.
+  """
+  bucket = storage.Client(project).get_bucket(bucket_name)
+  output_file_blob = bucket.blob(composite_name)
+  output_file_blob.content_type = 'text/plain'
+  blobs = [bucket.get_blob(blob_name) for blob_name in blob_names]
+  output_file_blob.compose(blobs)
+
+
+def _create_blob(client, file_path):
+  # type: (storage.Client, str) -> storage.Blob
+  bucket_name, blob_name = gcsio.parse_gcs_path(file_path)
+  file_blob = client.get_bucket(bucket_name).blob(blob_name)
+  file_blob.content_type = 'text/plain'
+  return file_blob
+
+
+class MultiProcessComposer(object):
+  """Class to compose (a large number of) files in GCS in parallel."""
+
+  _NUM_PROCS = 63
```

Comment:
Curious why you chose this number? Is it related to a GCS limit?

Reply:
I have done some experiments with this number, it seems that it is not necessary to specify this value.

Quote from [composite doc](https://cloud.google.com/storage/docs/composite-objects):
`There is a per-project rate limit on how many source objects you can compose of approximately 200 source objects per second.`

So I think roughly 6 processes can be created to run in parallel.

To compose 595 blobs, without this number, it takes roughly took about 20s. With `_NUM_PROCS = 6`, `_NUM_PROCS = 7`, `_NUM_PROCS = 16`, it took about the same time (25s). With `_NUM_PROCS = 1`, it took about 95 s. So I think it might be best to let the system decide the number of processes.

I will try to run with a bigger data sets to make sure it won't break.